pykokkos/lib/ufuncs.py

```diff
@@ -256,3 +256,491 @@ def sign(view):
+            power_impl_1d_double,
+            view=view,
+            viewA=viewA,
+            viewB=viewB)
+
+    elif str(viewA.dtype) == "DataType.float":
+        view = pk.View([viewA.shape[0]], pk.float)
+        pk.parallel_for(
+            viewA.shape[0],
+            power_impl_1d_float,
+            view=view,
+            viewA=viewA,
+            viewB=viewB)
+
+    return view
+
+
+@pk.workunit
+def fmod_impl_1d_double(tid: int, view: pk.View1D[pk.double], viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double]):
+    view[tid] = fmod(viewA[tid], viewB[tid])
+
+@pk.workunit
+def fmod_impl_1d_float(tid: int, view: pk.View1D[pk.float], viewA: pk.View1D[pk.float], viewB: pk.View1D[pk.float]):
+    view[tid] = fmod(viewA[tid], viewB[tid])
+
+
+def fmod(viewA, viewB):
+    """
+    Element-wise remainder of division when element of viewA is
+    divided by positionally corresponding element of viewB
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+    viewB : pykokkos view
+
+    Returns
+    -------
+    view : pykokkos view
+    """
+
+    if str(viewA.dtype) == "DataType.double":
+        view = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            fmod_impl_1d_double,
+            view=view,
+            viewA=viewA,
+            viewB=viewB)
+
+    elif str(viewA.dtype) == "DataType.float":
+        view = pk.View([viewA.shape[0]], pk.float)
+        pk.parallel_for(
+            viewA.shape[0],
+            fmod_impl_1d_float,
+            view=view,
+            viewA=viewA,
+            viewB=viewB)
+
+    return view
+
+
+@pk.workunit
+def square_impl_1d_double(tid: int, view: pk.View1D[pk.double]):
+    view[tid] = view[tid] * view[tid]
+
+@pk.workunit
+def square_impl_1d_float(tid: int, view: pk.View1D[pk.float]):
+    view[tid] = view[tid] * view[tid]
+
+def square(view):
+    """
+    Squares argument element-wise
+
+    Parameters
+    ----------
+    view : pykokkos view
+
+    Returns
+    -------
+    view : pykokkos view
+    """
+    if str(view.dtype) == "DataType.double":
+        pk.parallel_for(
+            view.shape[0],
+            square_impl_1d_double,
+            view=view)
+
+    elif str(view.dtype) == "DataType.float":
+        pk.parallel_for(
+            view.shape[0],
+            square_impl_1d_float,
+            view=view)
+
+    return view
+
+@pk.workunit
+def greater_impl_1d_double(tid: int, view: pk.View1D[pk.double], viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double]):
+    view[tid] = viewA[tid] > viewB[tid]
+
+@pk.workunit
+def greater_impl_1d_float(tid: int, view: pk.View1D[pk.float], viewA: pk.View1D[pk.float], viewB: pk.View1D[pk.float]):
+    view[tid] = viewA[tid] > viewB[tid]
+
+
+def greater(viewA, viewB):
+    """
+    Return the truth value of viewA > viewB element-wise.
+
+    Parameters
+    ----------
+    view : pykokkos view
+
+    Returns
+    -------
+    view : pykokkos view
```

Comment:
I think we should use a smaller datatype, but did we ever decide if we will be adding the `bool` datatype to `pykokkos-base`?

Reply:
> did we ever decide if we will be adding the bool datatype to pykokkos-base

I don't think we reached a final decision, but we probably should for array API conformance, rather than the current `bool` hacks we have over here. I think the NumPy `bool` type is a full byte per i.e., https://stackoverflow.com/questions/5602155/numpy-boolean-array-with-1-bit-entries.

Would we have a compelling reason to try to get away with less than 8 bits for our `bool`? Maybe we need an issue/discussion where we look at what NumPy vs. C++ do with `bool` arrays, and how hard it would be to get `repr` that shows `True` and `False` instead of `1` and `0` when you print, etc.

Reply:
Yeah we should definitely add `bool` when we get the chance. It would be nice to have less than 8 bits for `bool` but I don't think that's high priority for now.